src/Dfe.EarlyYearsQualification.Web/Program.cs

```diff
@@ -51,11 +52,24 @@
 builder.Services.AddTransient<ICookieService, CookieService>();
 
 builder.Services.AddSingleton<IActionContextAccessor, ActionContextAccessor>();
-builder.Services.AddScoped(x => {
-    var actionContext = x.GetRequiredService<IActionContextAccessor>().ActionContext;
-    var factory = x.GetRequiredService<IUrlHelperFactory>();
-    return factory.GetUrlHelper(actionContext!);
-});
+builder.Services.AddScoped(x =>
+                           {
+                               var actionContext = x.GetRequiredService<IActionContextAccessor>().ActionContext;
+                               var factory = x.GetRequiredService<IUrlHelperFactory>();
+                               return factory.GetUrlHelper(actionContext!);
+                           });
+
+var accessIsChallenged = !builder.Configuration.GetValue<bool>("ServiceAccess:IsPublic");
+// ...by default, challenge the user for the secret value unless that's explicitly turned off
+
+if (accessIsChallenged)
+{
+    builder.Services.AddScoped<IChallengeResourceFilterAttribute, ChallengeResourceFilterAttribute>();
+}
+else
+{
+    builder.Services.AddSingleton<IChallengeResourceFilterAttribute, NoChallengeResourceFilterAttribute>();

```

Comment:
Do we need this? If accessIsChallenged is false it wouldn't add a filter and would continue to work as normal?

Reply:
I understand why we need it now :-)

Reply:
Unfortunately we do. I needed to use dependency injection in the filter attribute: with that constraint I can either explicitly _always_ use the `ChallengeResourceFilterAttribute` which does more processing per request than necessary if access is public, or specify a no-op `IChallengeResourceFilterAttribute` for use when access is set to `IsPublic`. I also made sure that the `ChallengeResourceFilterAttribute` passes through if access is set to `IsPublic`, just to be sure.